bot/exts/valentines/markov_poem_generator.py

```diff
@@ -0,0 +1,506 @@
+                return await func(instance, ctx, scheme, unit_count,
+                                  time_start, is_first_error)
+        except TimeoutError:
+            await ctx.send(f"Sorry {ctx.author.mention}, but your poem"
+                           " timed out, please try again!")
+
+    return wrapper
+
+
+class MarkovPoemGenerator(commands.Cog):
+    """
+    A cog that provides a poem by taking the markov of a corpus.
+
+    By processing the corpus text through a markov chain, a series of lines
+    can be iterated through, whilst corresponding to the given rhyme scheme.
+    """
+
+    POEM_TIMEOUT = 60  # In seconds
+    near_rhyme_min_score = 2000
+    rhyming_line_finder_limiter = 80000
+    max_char_range = (50, 120)  # For the sentence generator
+
+    SOURCES: List[str] = [
+        "shakespeare_corpus.txt"
+    ]
+
+    templates: Dict[str, str] = {
+        "shakespearean-sonnet": "abab/cdcd/efef/gg",
+        "spenserian-sonnet": "abab/bcbc/cdcd/ee",
+        "petrarch-sonnet": "abbaabba/cdecde",
+        "ballade": "ababbcbc",
+        "terza-rima": "aba/bcb/cdc/ded/ee",
+        "villanelle": "aba/aba/aba/aba/aba/abaa",
+        "limerick": "aabba"
+    }
+
+    rhyme_websites: List[Tuple[bool, str]] = [
+        # (is exact rhyme, website link)
+        (True, "https://api.datamuse.com/words?rel_rhy="),
+        (False, "https://api.datamuse.com/words?rel_nry=")
+    ]
+
+    def __init__(self, bot: commands.Bot):
+        """Initializes the full corpus text and the markov model."""
+        self.bot = bot
+
+        # Load the full text corpus
+        for source_file in self.SOURCES:
+            with Path(f"bot/resources/valentines/{source_file}").open() as f:
+                full_corpus = f.read().splitlines()
+
+        # Create the markov model
+        self.model = markovify.Text(full_corpus, state_size=1)
+
+        logging.info("Full text corpus and markov model successfully loaded.")
+
+    @staticmethod
+    def _get_unit_count(scheme: str) -> Dict[str, bool]:
+        """Checks how many times a unit occurs in a rhyme scheme."""
+        return {char: scheme.count(char) for char in set(scheme)}
+
+    @staticmethod
+    def _get_last_word(sentence: str) -> str:
+        """Returns the last word of a sentence string."""
+        return sentence.strip(string.punctuation).split()[-1]
+
+    async def _get_sentence(self) -> str:
+        """
+        Uses `in_thread` to return a sentence asynchronously.
+
+        The sentence is generated via the markov chain.
+        """
+        def func() -> str:
+            line = self.model.make_short_sentence(
+                random.randint(*self.max_char_range)
+            )
+
+            if not isinstance(line, str):
+                logging.error(f"Argument is a {type(line)} instead of a str.")
+                raise MakeShortSentenceRanOut
+
+            return line
+
+        return await in_thread(func)
```

Comment:
I suppose I could replace the return statement with
`return_value = await in_thread(func)`
`return return_value`

but I feel like that's redundant.

Making the return type hint a `Future` instead of `Str` makes the type hint less useful. What do you suggest?

Reply:
We can ignore this ig.